core/graph/src/main/java/uk/gov/gchq/gaffer/graph/hook/migrate/SchemaMigration.java

```diff
@@ -0,0 +1,645 @@
+/*
+ * Copyright 2018 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.gchq.gaffer.graph.hook.migrate;
+
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import org.apache.commons.collections.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import uk.gov.gchq.gaffer.data.element.ElementTuple;
+import uk.gov.gchq.gaffer.data.element.function.ElementFilter;
+import uk.gov.gchq.gaffer.data.element.function.ElementTransformer;
+import uk.gov.gchq.gaffer.data.elementdefinition.view.View;
+import uk.gov.gchq.gaffer.data.elementdefinition.view.ViewElementDefinition;
+import uk.gov.gchq.gaffer.graph.hook.AddOperationsToChain;
+import uk.gov.gchq.gaffer.graph.hook.GraphHook;
+import uk.gov.gchq.gaffer.operation.Operation;
+import uk.gov.gchq.gaffer.operation.OperationChain;
+import uk.gov.gchq.gaffer.operation.graph.OperationView;
+import uk.gov.gchq.gaffer.operation.impl.function.Aggregate;
+import uk.gov.gchq.gaffer.operation.impl.function.Filter;
+import uk.gov.gchq.gaffer.operation.impl.function.Transform;
+import uk.gov.gchq.gaffer.store.Context;
+import uk.gov.gchq.gaffer.store.schema.ViewValidator;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link SchemaMigration} {@link GraphHook} allows an admin to set migration mappings
+ * that are then applied on any {@link Operation} with output and an {@link OperationView}.
+ * <p>
+ * To make use of this {@link SchemaMigration} the implemented {@link uk.gov.gchq.gaffer.store.Store} must have the Transform trait.
+ */
+@JsonPropertyOrder(value = {"entities", "edges", "transformToNew"}, alphabetic = true)
+public class SchemaMigration implements GraphHook {
+    public static final MigrationOutputType DEFAULT_OUTPUT_TYPE = MigrationOutputType.OLD;
+    public boolean aggregateAfter = false;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SchemaMigration.class);
+
+    private List<MigrateElement> entities = new ArrayList<>();
+    private List<MigrateElement> edges = new ArrayList<>();
+
+    private MigrationOutputType outputType = DEFAULT_OUTPUT_TYPE;
+
+    private Map<String, ElementFilter> entitiesPostAggregationFilterMap = new HashMap<>();
```

Comment:
Graph hooks need to be stateless, because the same instance is used for every query for different users. So we can't store these maps as fields, they will need to be local variables.